interactive_engine/compiler/src/main/java/com/alibaba/graphscope/common/ir/planner/GraphFieldTrimmer.java

```diff
@@ -0,0 +1,214 @@
+import com.alibaba.graphscope.common.ir.rel.type.group.GraphGroupKeys;
+import com.alibaba.graphscope.common.ir.rex.RexGraphVariable;
+import com.alibaba.graphscope.common.ir.rex.RexVariableAliasCollector;
+import com.alibaba.graphscope.common.ir.tools.GraphBuilder;
+import com.alibaba.graphscope.common.ir.type.GraphProperty;
+import com.google.common.collect.ImmutableSet;
+import org.apache.calcite.plan.hep.HepProgram;
+import org.apache.calcite.rel.AbstractRelNode;
+import org.apache.calcite.rel.RelCollation;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.logical.LogicalFilter;
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.sql.validate.SqlValidator;
+import org.apache.calcite.sql2rel.RelFieldTrimmer;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.util.ReflectUtil;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class GraphFieldTrimmer extends RelFieldTrimmer {
+    private GraphBuilder graphBuilder;
+
+    public class GraphVariable {
+        private int aliasId;
+        private @Nullable GraphProperty property;
+
+        public GraphVariable(int aliasId, @Nullable GraphProperty property) {
+            this.aliasId = aliasId;
+            this.property = property;
+        }
+
+        public GraphVariable(int aliasId) {
+            this.aliasId = aliasId;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            if (!super.equals(o)) return false;
+            RexGraphVariable rhs = (RexGraphVariable) o;
+            if (rhs.getProperty() == null) {
+                return aliasId == rhs.getAliasId();
+            }
+            return aliasId == rhs.getAliasId() && Objects.equals(property, rhs.getProperty());
+
+        }
+    }
+
+    private final ReflectUtil.MethodDispatcher<RelNode> graphTrimFieldsDispatcher;
+
+    public GraphFieldTrimmer(GraphBuilder builder) {
+        super(null, builder);
+        graphBuilder = builder;
+        graphTrimFieldsDispatcher =
+                ReflectUtil.createMethodDispatcher(
+                        RelNode.class,
+                        this,
+                        "trimFields",
+                        RelNode.class,
+                        Set.class);
+
+
+    }
+
+    public RelNode trim(RelNode root) {
+        final ImmutableSet<GraphVariable> fields = ImmutableSet.of();
+        return dispatchTrimFields(root, fields);
+    }
+
+
+    public RelNode trimFields(GraphLogicalProject project, ImmutableSet<GraphVariable> fieldsUsed) {
+        // TODO(huaiyu)
+        return project;
+    }
+
+    public RelNode trimFields(GraphLogicalAggregate aggregate, ImmutableSet<GraphVariable> fieldsUsed) {
```

Comment:
There's no need to implement a TrimField function for every operator, the difference only occurs when collecting the variables used by the operator. The rest of the process can be unified.